tests/integration/test_upgrade.py

```diff
@@ -0,0 +1,257 @@
+import subprocess
+import time
+
+import kubernetes
+import ops.model
+import PIL.Image
+import pytest
+import pytest_operator.plugin
+import requests
+from playwright.async_api import async_playwright
+
+from tests.integration.wordpress_client_for_test import WordpressClient
+
+logger = logging.getLogger()
+
+POST_CONTENT = """\
+WordPress powers more than 39% of the web — a figure that rises every day.
+Everything from simple websites, to blogs, to complex portals and enterprise
+websites, and even applications, are built with WordPress. WordPress combines
+simplicity for users and publishers with under-the-hood complexity for
+developers. This makes it flexible while still being easy-to-use."""
+
+POST_CONTENT = POST_CONTENT.replace("\n", " ")
+
+POST_TITLE = "WordPress Post #1"
+
+POST_COMMENT = "I am a comment."
+
+
+async def screenshot(url, path):
+    """Create a screenshot of a website."""
+    async with async_playwright() as p:
+        browser = await p.chromium.launch()
+        page = await browser.new_page()
+        await page.goto(url)
+        time.sleep(10)
+        await page.screenshot(full_page=True, path=path)
+        await browser.close()
+
+
+@pytest.fixture(scope="module", name="gen_upgrade_test_charm_config")
+def gen_upgrade_test_charm_config_fixture(ops_test, swift_config, kube_core_client):
+    """Create a function that generates charm config for upgrading tests."""
+    swift_url = swift_config["swift-url"]
+    swift_config = copy.copy(swift_config)
+    swift_config["url"] = f"{swift_url}/{swift_config['bucket']}/wp-content/uploads/"
+    del swift_config["swift-url"]
+    swift_config["prefix"] = swift_config["object-prefix"]
+    del swift_config["object-prefix"]
+
+    def _gen_upgrade_test_charm_config():
+        charm_config = {
+            "db_host": kube_core_client.read_namespaced_pod(
+                name="mysql", namespace=ops_test.model_name
+            ).status.pod_ip,
+            "db_name": "wordpress",
+            "db_user": "wordpress",
+            "db_password": "wordpress-password",
+        }
+        if swift_config:
+            logger.info(f"deploy with swift: {swift_config}")
+            charm_config["wp_plugin_openstack-objectstorage_config"] = json.dumps(swift_config)
+        return charm_config
+
+    return _gen_upgrade_test_charm_config
+
+
+@pytest.mark.abort_on_fail
+async def test_deploy_old_version(
+    num_units,
+    ops_test: pytest_operator.plugin.OpsTest,
+    application_name,
+    kube_core_client,
+    deploy_and_wait_for_mysql_pod,
+    gen_upgrade_test_charm_config,
+):
+    """
+    arrange: none.
+    act: deploy all required charms and kubernetes pods for tests.
+    assert: all charms and pods are deployed successfully.
+    """
+    assert ops_test.model
+
+    async def deploy_wordpress():
+        await ops_test.model.deploy(
+            "wordpress-k8s",
+            resources={"wordpress-image": "wordpresscharmers/wordpress:v5.9.4-20.04_edge"},
```

Comment:
Why this version explicitly ?
This whole test seems to cover this very specific case, do we want to keep such specific test in the repo ?

Reply:
It's the latest version of the CharmHub version of the WordPress charm supported, and the new sidecar version is based on that. The entire `test_upgrade` is to mimic upgrading from one of our production deployment of the old WordPress charm. So some config is picked from the production config.

Reply:
Ok let's keep it like this, but if in the future we want to test other version upgrade, we'll probably need to find a way to generalize these tests